zebrad/src/components/mempool.rs

```diff
@@ -76,6 +78,12 @@ pub struct Mempool {
 
     /// The transaction dowload and verify stream.
     tx_downloads: Pin<Box<InboundTxDownloads>>,
+
+    /// Allows checking if we are near the tip to enable/disable the mempool.
+    sync_status: SyncStatus,
+
+    /// Indicates whether the mempool is enabled or not.
+    enabled: bool,
```

Comment:
I want to find a better way to represent this state of the mempool. Some ideas: 
- `state` is an enum variant like `State::{Enabled, Disabled}`
- Make `Mempool` itself an enum with variants like above, that still impls `Service`, etc, to model `Mempool` itself as a state machine. An example of this is in https://github.com/ZcashFoundation/zebra/blob/9b3f6b3105263377e6836ff93a0f25f273a9dfd9/zebra-network/src/peer/connection.rs

@jvff thoughts?

Reply:
I like the second one, maybe not the Mempool itself but a field, like the `Setup` enum in the `Inbound` service https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/components/inbound.rs#L44

Then we could move the storage and downloader into the `Enabled` state, automatically destroying them when the mempool is deactivated.  This would fall more into the scope of https://github.com/ZcashFoundation/zebra/issues/2723, though. Maybe I can just remove the `disabled` flag here and this can be handled by #2723?

(But yeah, let's wait for @jvff :sweat_smile: )

Reply:
I think any of the alternatives are fine :sweat_smile: 

1. `bool` is simple enough for now, but might be a code smell and future work might lead to `Mempool` representing an invalid state.
2. A simple `enum State` just adds semantics to the `bool`, but it also offers the possibility of attaching things to the state in the future.
3. Making the whole `Mempool` an `enum` is a more robust solution, with the caveat that it may add more complexity and verbosity (i.e., may require multiple methods to start with `match self`), especially if there are fields that should be common among the variants (e.g., the `sync_states` comes to mind).

I think I'd prefer a separate `enum State` (option number 2), that could also help with clearing the mempool and stopping the downloader when the mempool becomes disabled. Drafting a rough idea here, bear with me:

```
enum State {
    Disabled,
    Enabled {
        storage: storage::Storage,
        tx_downloads: Pin<Box<InboundTxDownloads>>,
    },
}
```

This would make the type system ensure that the downloader has stopped and that the storage is cleared when the mempool is disabled. But I guess I'd start with empty variants first, and move the fields from `Mempool` to `mempool::State` in separate PRs.

Again, just an idea (and a rough sketch of what came to my mind), and I'm fine with any of the options :sweat_smile:

Reply:
One thing I want to nail down before we choose is what happens when we turn mempool on and off, like if we thought we were near the chain tip, turned on mempool, then started syncing down a bunch of blocks, turn it off again, etc.

Do we throw away all state, such as the verified set, the download and verify queue, the crawlers, and importantly, the rejected set? Should we throw away everything but the rejected set? (I'm thinking about some of the ZIP-401 requirements as I ask this).

I think it would be _ok_ to throw everything away on state change for now. If that results in thrashing because our on/off triggers are finicky, we may revisit the trigger or the state tossing, but since that may be a pre-mature optimization, I like being able to throw away all the above statefulness on enabled/disabled change.

So, yeah, what @conradoplg said :
> Then we could move the storage and downloader into the Enabled state, automatically destroying them when the mempool is deactivated. This would fall more into the scope of #2723, though. Maybe I can just remove the enabled flag here and this can be handled by #2723?

especially as @jvff pointed out the `SyncStatues` needs not to be tossed:

>  it may add more complexity and verbosity (i.e., may require multiple methods to start with match self), especially if there are fields that should be common among the variants (e.g., the sync_states comes to mind).

So yeah, I would be cool removing the `enabled` bool in this PR and using this approach in #2723 🎉

Reply:
Removed `enabled` in 319fa852